deploy/multideploy.sh

```diff
@@ -0,0 +1,267 @@
+  _services=$(yq e '.services[].name' "$_deploy_file")
+  _debug2 "Services" "$_services"
+
+  if [ -z "$_services" ]; then
+    _err "Config does not have any services to deploy to."
+    return 1
+  fi
+  _debug2 "Config has services."
+
+  IFS=$(printf '\n')
+  # Check if extracted services exist in services list
+  for _service in $_services; do
+    _debug2 "Checking service" "$_service"
+    # Check if service exists
+    if ! yq e ".services[] | select(.name == \"$_service\")" "$_deploy_file" >/dev/null; then
+      _err "Service '$_service' not found."
+      return 1
+    fi
+
+    # Check if service has hook
+    if ! yq e ".services[] | select(.name == \"$_service\").hook" "$_deploy_file" >/dev/null; then
+      _err "Service '$_service' does not have a hook."
+      return 1
+    fi
+
+    # Check if service has environment
+    if ! yq e ".services[] | select(.name == \"$_service\").environment" "$_deploy_file" >/dev/null; then
+      _err "Service '$_service' does not have an environment."
+      return 1
+    fi
+  done
+  IFS=$OLDIFS
+}
+
+# Description: This function takes a list of environment variables in YAML format,
+#              parses them, and exports each key-value pair as environment variables.
+# Arguments:
+#   $1 - A string containing the list of environment variables in YAML format.
+# Usage:
+#   _export_envs "$env_list"
+_export_envs() {
+  _env_list="$1"
+
+  _secure_debug3 "Exporting envs" "$_env_list"
+
+  IFS=$(printf '\n')
+  echo "$_env_list" | yq e -r 'to_entries | .[] | .key + "=" + .value' | while IFS='=' read -r _key _value; do
+    _value=$(eval echo "$_value")
+    _savedomainconf "$_key" "$_value"
+    _secure_debug3 "Saved $_key" "$_value"
+  done
+  IFS=$OLDIFS
+}
+
+# Description:
+#   This function takes a YAML formatted string of environment variables, parses it,
+#   and clears each environment variable. It logs the process of clearing each variable.
+# Arguments:
+#   $1 - A YAML formatted string containing environment variable key-value pairs.
+# Usage:
+#   _clear_envs "<yaml_string>"
+_clear_envs() {
+  _env_list="$1"
+
+  _secure_debug3 "Clearing envs" "$_env_list"
+  env_pairs=$(echo "$_env_list" | yq e -r 'to_entries | .[] | .key + "=" + .value')
+
+  IFS=$(printf '\n')
+  echo "$env_pairs" | while IFS='=' read -r _key _value; do
+    _debug3 "Deleting key" "$_key"
+    _cleardomainconf "SAVED_$_key"
+    unset -v "$_key"
+  done
+  IFS="$OLDIFS"
+}
+
+# Description:
+#   This function deploys services listed in the deploy configuration file.
+# Arguments:
+#   $1 - The path to the deploy configuration file.
+#   $2 - The list of services to deploy.
+# Usage:
+#   _deploy_services "<deploy_file_path>" "<services_list>"
+_deploy_services() {
+  _deploy_file="$1"
+  _debug3 "Deploy file" "$_deploy_file"
+
+  _services=$(yq e '.services[].name' "$_deploy_file")
+  _debug3 "Services" "$_services"
+
+  printf '%s\n' "$_services" | while IFS= read -r _service; do
```

Comment:
```suggestion
_service_array=()
printf '%s\n' "$_services" | while IFS= read -r _service; do
  _service_array+=("$_service")
done

for _service in "${_service_array[@]}"; do
```

I think there is an issue caused by the ssh hook. The current ssh hook does not use the -n option, so it consumes stdin. This would cause the "read -r" to end.

Although I believe the default command for the ssh hook should be modified (by adding -n). But considering that the ssh hook saves the commands used, there might be users who have already used the ssh hook before, so it's best to read all the services first, I think.

Reply:
What issue did you experience with the ssh hook while using multideploy?

Reply:
It's possible, similar problem is with routeros hook. ssh and scp utils in this hooks read stdin and this breaks while statement after first hook with ssh or scp.

Reply:
> What issue did you experience with the ssh hook while using multideploy?

You can test the ssh hook along with other hooks, and you will find that the hooks after the ssh hook will not be processed because read -r has already been closed. As @lojzik said, it seems that this issue is not limited to just the ssh hook.